### DISCUSSION TITLE
How to use sequence, or custom vars on Prisma?

### DISCUSSION THREAD

**Discussion by darr1s** [factory-js/factory-js#28] [Q&A] [answered]
Hi there! Thanks for the great repo.

I am currently using this in my project and have noticed a few issues:

1. Prisma generated factories cannot create custom variables.
2. No sequence number accessor.
3. UUIDs are generated as strings, and createdAt fields with @default(now()) are also generated (as mentioned in the original README).

Given the current options of after + traits, I'm unsure if I should create all the factories. So, I wrote a script to generate base factories with ((id) => uuid()) based on the generated ones. I then use these base factories for my real e2e factories, making the seeders simpler and easier to manage than maintaining 100+ factories for schema models.

**Comment by ktmouk** [comment, marked as answer]
Hi, @darr1s. 👋

> 1. Prisma generated factories cannot create custom variables.

It is impossible to add new variables to the defined factory because of TypeScript limitation, but you can create a new factory with the defined factory by using [.def](https://github.com/factory-js/factory-js?tab=readme-ov-file#def), like below:

```typescript
const baseUserFactory = defineUserFactory(db)
const { props, vars } = baseUserFactory.def

export const userFactory = await factory
  .define({
    props: {
      ...props,
      foo: () => "foo",　// add 'foo' prop
    },
    vars: {
      ...vars,
      bar: () => "bar",　// add 'bar' var
    },
  })
```

> 3. UUIDs are generated as strings, and createdAt fields with @default(now()) are also generated (as mentioned in the original README).

Yes, prisma-factory currently does not support generating default values based on [native database type attributes](https://www.prisma.io/docs/orm/prisma-schema/data-model/models#native-types-mapping) because DMMF, used in prisma-factory, does not support passing native types (ref: https://github.com/prisma/prisma/issues/10252). Therefore, one simple way is to override the id property yourself, as you already described.

> 2. No sequence number accessor.

Could you please tell me more about the sequence number accessor? Is it similar to [sequences in factory_bot](https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#sequences)?

**Reply by darr1s** [reply]
Hi @ktmouk ,

Thanks for the prompt reply!

For sequence, yes it is a sequence number like factory_bot, useful when we are using `.buildList()` or `.createList()`

**Reply by ktmouk** [reply]
You can implement sequence generator by writing simple closure.

```typescript
const seq = <O>(iteratee: (count: number) => Promise<O> | O) => {
  let count = 0;
  return () => iteratee(count++);
};

export const userFactory = await factory
  .define({
    props: {
      name: seq(0, (n) => `name-${n}`)
    },
    vars: {}
  })
```
Actually, factory-js had the seq feature, which is the same as factory_bot, but we removed it at #17. This feature was intended to generate unique values and avoid UNIQUE constraint errors. However, since Jest/Vitest [runs tests in multiple threads by default](https://vitest.dev/guide/features#threads), it could not generate unique values across threads. But if you just want to generate sequence numbers, I think the above function would be helpful for you.

---

FYI: If the table colums has the [@default attribute](https://www.prisma.io/docs/orm/prisma-schema/data-model/models#defining-a-default-value), you can set `undefined` to props to use the default value that Prisma generates.


```typescript
// schema.schema
model User {
  id String @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
  createdAt DateTime @default(now()) @db.Timestamptz(3)
}

// factory.ts
export const userFactory = defineUserFactory(db).props({
  id: () => undefined,  
  createdAt: () => undefined
})
```
